### DISCUSSION TITLE
any(), all() functions

### DISCUSSION THREAD

**Discussion by innovate-invent** [jmespath-community/jmespath.spec#42] [Ideas]
It may be useful to include basic boolean functions:

`boolean all(object|array $arg)` - Returns true if all values are truthy, false otherwise
`boolean any(object|array $arg)` - Returns true if any values are truthy, false otherwise

**Comment by springcomp** [comment]
Maybe extend this with an extra `boolean $predicate` argument?

```
search({"a": [1, 2], "b": [3, 4]}, any(@, &contains(2))) -> true
search({"a": [1, 2], "b": [3, 4]}, all(@, &contains(2))) -> false
```

**Reply by innovate-invent** [reply]
I think that rather than a predicate, we could just use a projection: `any(*.contains(@, 2))`

**Reply by springcomp** [reply]
Good catch.
By the way, the `contains(@, 2)` currently does not work. We may want to extend `contains()` to accept values instead of just strings.

**Reply by innovate-invent** [reply]
does not work? I just tested it in the demo
did you add the backticks around the literal? ``contains(@, `2`)``

**Reply by springcomp** [reply]
You are right, it does. My mistake.

**Comment by innovate-invent** [comment]
I would forgo these functions in place of the reduce operation being proposed combined with the `&&` and `||` operators

**Reply by springcomp** [reply]
Can you share the resulting expressions that this would entails ?

**Reply by innovate-invent** [reply]
all: ``search([&`true`].acc && cur, [true, true, true])``
any: ``search([&`true`].acc || cur, [true, true, true])``
